loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+
+```java
+/**
+* @param <T> the types of resolver this plugin can resolve
+*/
+interface QuiltLoaderPlugin<T extends CandidateResolver> extends  {
+/**
+* Called once per cycle as the first action in the cycle.
+* 
+* This is where mods can be added with {@link QuiltPluginContext#addCandidate} and
+* {@link QuiltPluginContext#addTentativeCandidate}.
+*/
+default void run(QuiltPluginContext context) {}
+
+/**
+* Called once per cycle after the sat4j solving has finished, but before any resolvers are run.
+*
+* Should NOT invoke the resolvers.
+*
+* @return true if all of the resolvers can be called, false otherwise
+*/
+default boolean canResolve(List<T> resolvers) {
+    return false;
+}
+
+/**
+* Called if loader can't simplify this error down into any of the other error handling methods.
+* @return True if this plugin did something which will solve / change the error in future,
+*         and so loader won't ask any other plugins to solve this.
+*         Loader will temporarily remove this rule so it won't be sent to #handleOtherErrors again in this cycle.
+*         If this returns false then no rules will be removed, and instead loader will assume that
+*         the error has been handled in some other way. (and it will promptly crash if you haven't)
+*/
+default @Nullable Rule handleOtherErrors(QuiltPluginContext context, List<Rule> errorChain) { return null; }
+
+/**
+* @param dep The dependency which is missing completely. If you can find a valid source for this then you should add 
+*            it with {@link QuiltContext#addTentativeCandidate()}
+* @return True if this plugin did something which will solve / change the error in future,
+*         and so loader won't ask any other plugins to solve this.
+*         Loader will temporarily remove this rule so it won't be sent to #handleOtherErrors again in this cycle.
+*         If this returns false then no rules will be removed, and instead loader will assume that
+*         the error has been handled in some other way. (and it will promptly crash if you haven't)
+*/
+default boolean handleMissingDependencyError(QuiltPluginContext context, ModDependencyRule dep, List<Rule> fullErrorChain) {
+    return handleOtherErrors(ctx, fullErrorChain);
+}
+
+/**
+* Called whenever a new LoadOption is added, for plugins to add Rules based on this. (For example the default plugin 
+* creates rules based on the dependencies and breaks sections of the quilt.mod.json if this option is a 
+* {@link MainModLoadOption}).
+* <p>
+* Most plugins are not expected to implement this.
+*/
+default void onLoadOptionAdded(QuiltPluginContext context, LoadOption option) {}
+}
+```
+</details>
+
+<details>
+    <summary>ITentativeLoadOption</summary>
+    
+```java
+/**
+* {@link LoadOption}s can implement this if they must be processed at the end of the cycle in order to either be
+* added as a normal LoadOption, or removed automatically.
+*/
+public interface ITentativeLoadOption {
```

Comment:
Nit: `I` class prefix doesn't align with what I'd say are the agreed but not concrete naming standards at Quilt

Reply:
These were the names Alex gave me, and they work well enough to convey the meaning for now. Very few if any of them are likely to make it into a final implementation 😄